### ISSUE TITLE
set_data_arg option throwing (unnecessary) warning

### ISSUE THREAD

**Issue by jwilliman** [singmann/afex#49]
When running `mixed` (afex 0.19.1) the option `set_data_arg` throws a warning despite working as intended.

```
data(obk.long)
m <- mixed(value ~ treatment + (1|id), obk.long, set_data_arg = TRUE)
Contrasts set to contr.sum for the following variables: treatment, id
Fitting one lmer() model. [DONE]
Calculating p-values. [DONE]
Warning messages:
1: extra argument(s) ‘set_data_arg’ disregarded 
2: extra argument(s) ‘set_data_arg’ disregarded
> m$full_model@call
lme4::lmer(formula = value ~ treatment + (1 | id), data = obk.long, 
    set_data_arg = TRUE)

> m <- mixed(value ~ treatment + (1|id), obk.long, set_data_arg = FALSE)
Contrasts set to contr.sum for the following variables: treatment, id
Fitting one lmer() model. [DONE]
Calculating p-values. [DONE]
Warning messages:
1: extra argument(s) ‘set_data_arg’ disregarded 
2: extra argument(s) ‘set_data_arg’ disregarded 
> m$full_model@call
lme4::lmer(formula = value ~ treatment + (1 | id), data = data, 
    set_data_arg = FALSE)

```

Also I'm somewhat confused about the way this works inside a function. It appears to do the opposite of what I expect, by only working with `emmeans` when set to `FALSE`?

```

f <- function(dat, set_data_arg) {
  m <- mixed(value ~ treatment + (1|id), dat, set_data_arg = set_data_arg)
  emmeans(m, "treatment")  
}

> f(obk.long, TRUE)
Contrasts set to contr.sum for the following variables: treatment, id
Fitting one lmer() model. [DONE]
Calculating p-values. [DONE]
Error in is.data.frame(data) : object 'dat' not found
In addition: Warning messages:
1: extra argument(s) ‘set_data_arg’ disregarded 
2: extra argument(s) ‘set_data_arg’ disregarded 
 Error in ref_grid(object, ...) : 
  Perhaps a 'data' or 'params' argument is needed 

> f(obk.long, FALSE)
Contrasts set to contr.sum for the following variables: treatment, id
Fitting one lmer() model. [DONE]
Calculating p-values. [DONE]
 treatment emmean        SE df lower.CL upper.CL
 control     4.20 0.6536551 13 2.787864 5.612136
 A           6.25 0.7308086 13 4.671184 7.828816
 B           6.00 0.5524394 13 4.806527 7.193473

Degrees-of-freedom method: kenward-roger 
Confidence level used: 0.95 
Warning messages:
1: extra argument(s) ‘set_data_arg’ disregarded 
2: extra argument(s) ‘set_data_arg’ disregarded

```

**Comment by singmann** [comment]
Thanks a lot for this detailed issue and I apologize for my delay in reacting.

The first problem is simply a bug. `set_data_arg` should not throw a warning and does not anymore now in the development version (fixed in e8f3f0f):

```
data(obk.long)
m <- mixed(value ~ treatment + (1|id), obk.long, set_data_arg = TRUE)
# Contrasts set to contr.sum for the following variables: treatment, id
# Fitting one lmer() model. [DONE]
# Calculating p-values. [DONE]
```

The other issue is really surprising. One could have expected something similar as you expected it. But then, using functions that work with `model.frame` and stuff like that in other functions is always a recipe for a lot of surprises. However, I have tracked down the reason why it works with `set_data_arg = FALSE`. (As I said, one could expect it to work with `set_data_arg = TRUE`. But the fact that it does not is also within the realm of possibilities.)

The reason it works if `set_data_arg = FALSE` is the following. The function that is called by `emmeans` and does the recovering is `emmeans:::recover_data.call`. It not only receives the object (i.e.,t he fitted `lmer` model) but the `terms()`. And these terms of the fitted `lmer` model contain an environment that seems to contain all the objects used within the `mixed` functions. One of those the `data` argument used by `mixed`. This was used for fitting and is now found:

```
env <- environment(terms(m$full_model))
ls.str(envir = env)
# all.terms :  chr [1:2] "treatment" "1 | id"
# all_fit :  logi FALSE
# anova_out : Classes ‘anova’ and 'data.frame':	1 obs. of  6 variables:
#  $ Sum Sq : num 16.9
#  $ Mean Sq: num 8.46
#  $ NumDF  : int 2
#  $ DenDF  : num 13
#  $ F.value: num 2.91
#  $ Pr(>F) : num 0.09
# anova_tab_addition :  NULL
# anova_table : Classes ‘anova’ and 'data.frame':	1 obs. of  4 variables:
#  $ num Df: int 2
#  $ den Df: num 13
#  $ F     : num 2.91
#  $ Pr(>F): num 0.09
# args_test :  list()
# c.ns :  chr(0) 
# check_contrasts :  logi TRUE
# cl :  NULL
# data : 'data.frame':	240 obs. of  7 variables:
#  $ id       : Factor w/ 16 levels "1","2","3","4",..: 1 1 1 1 1 1 1 1 1 1 ...
#  $ treatment: Factor w/ 3 levels "control","A",..: 1 1 1 1 1 1 1 1 1 1 ...
#  $ gender   : Factor w/ 2 levels "F","M": 2 2 2 2 2 2 2 2 2 2 ...
#  $ age      : num  -4.75 -4.75 -4.75 -4.75 -4.75 -4.75 -4.75 -4.75 -4.75 -4.75 ...
#  $ phase    : Factor w/ 3 levels "fup","post","pre": 3 3 3 3 3 2 2 2 2 2 ...
#  $ hour     : Factor w/ 5 levels "1","2","3","4",..: 1 2 3 4 5 1 2 3 4 5 ...
#  $ value    : num  1 2 4 2 1 3 2 5 3 2 ...
# dots :  list()
# dv :  chr "value"
# effect.order :  int 1
# expand_re :  logi FALSE
# fits :  NULL
# fixed.effects :  chr "treatment"
# fixed.vars :  chr "treatment"
# formula : Class 'formula'  language value ~ treatment + (1 | id)
# formula.f : Class 'formula'  language value ~ treatment + (1 | id)
# full_model : Formal class 'merModLmerTest' [package "lmerTest"] with 13 slots
# i :  chr "id"
# list.out : List of 5
#  $ anova_table      :Classes ‘anova’ and 'data.frame':	1 obs. of  4 variables:
#  $ full_model       :Formal class 'merModLmerTest' [package "lmerTest"] with 13 slots
#  $ restricted_models: NULL
#  $ tests            : NULL
#  $ data             :'data.frame':	240 obs. of  7 variables:
# m.matrix :  num [1:240, 1:3] 1 1 1 1 1 1 1 1 1 1 ...
# mapping :  int [1:3] 0 1 1
# max.effect.order :  int 1
# mc :  language mixed(formula = value ~ treatment + (1 | id), data = obk.long, set_data_arg = TRUE)
# method :  chr "KR"
# mf :  language lmerTest::lmer(formula = value ~ treatment + (1 | id), data = data)
# per_parameter :  NULL
# progress :  logi TRUE
# random :  chr "(1 | id)"
# resetted :  chr [1:2] "treatment" "id"
# return :  chr "mixed"
# rh2 : Class 'formula'  language ~treatment
# set_data_arg :  logi TRUE
# sig_symbols :  chr [1:4] " +" " *" " **" " ***"
# test_intercept :  logi FALSE
# tests :  NULL
# type :  num 3
# use_reml :  logi TRUE
# vars.to.check :  chr [1:3] "value" "treatment" "id"
```

I am not sure what to take from this. Maybe it would be better to make the default for `set_data_arg = FALSE` as this leads to less errors? I can not fully foresee what happens then. I am happy for any comments.

**Comment by singmann** [comment]
`set_data_arg` can now be set via `afex_options`. See `afex_options("set_data_arg")`. The default is:

```
> afex_options("set_data_arg")
[1] FALSE
```

I think this solves the problem.
